### DISCUSSION TITLE
Invalid usage of optional parameter / missing usage of Optional types

### DISCUSSION THREAD

**Discussion by dctr** [encode/httpx#1747] [Potential Issue] [answered]
A lot of parameters in httpx are provided a fallback-value of `None`, whilst the respective types are not indicating being optional.

Example from https://github.com/encode/httpx/blob/master/httpx/_client.py

```python
    async def request(
        self,
        method: str,
        url: URLTypes,
        *,
        content: RequestContent = None,
        data: RequestData = None,
        files: RequestFiles = None,
        json: typing.Any = None,
        params: QueryParamTypes = None,
        headers: HeaderTypes = None,
        cookies: CookieTypes = None,
        auth: typing.Union[AuthTypes, UnsetType] = UNSET,
        allow_redirects: bool = True,
        timeout: typing.Union[TimeoutTypes, UnsetType] = UNSET,
    ) -> Response:
        # ...
```

with for example

```python
RequestContent = Union[str, bytes, Iterable[bytes], AsyncIterable[bytes]]
```

It should either be

```python
# parameter
content: Optional[RequestContent] = None
# or type
RequestContent = Union[str, bytes, Iterable[bytes], AsyncIterable[bytes], None]  # including None
```

The latter would align with what is done e.g. in https://github.com/encode/httpx/blob/master/httpx/_types.py with

```python
AuthTypes = Union[
    Tuple[Union[str, bytes], Union[str, bytes]],
    Callable[["Request"], "Request"],
    "Auth",
    None,
]
```

From personal preference I'd be in favour of the former though, as it makes the type-signature more meaningful.

For context, I currently have a wrapper function around `request`, but have to silence type warnings.

**Comment by FasterSpeeding** [comment, marked as answer]
#1439 seems related to this but as random input i've also had a similar issue with having to silence type warnings when wrapping the AsyncClient

**Comment by lovelydinosaur** [comment]
I mean it's a real mixed bag. Yeah we *could* but it'd add an awful lot of extra noise to the codebase.

I *guess* I'd be happy to review how the change looked if someone wanted to go ahead and issue a pull request that changed all those signatures, and switched up our own mypy rules to enforce it, but I certainly wouldn't be able to say if we'd accept the pull request or not, so it'd need someone who's willing to put in that time, without having any guarantees that it'd make it in.

**Reply by altendky** [reply]
My perspective is that this is less about "is `Optional[]` ugly" and more about whether or not httpx wants to enable their users to have stricter hint checking or not.  This is a feature and `Optional[]` is how it is implemented.

It would be nice for mypy to automatically use the library's mypy configuration but until they do we're stuck with this.  Adding `Optional[]` everywhere will undoubtedly be ugly if you prefer the implicit approach.  It seems unlikely that seeing the ugliness would convince anyone it isn't ugly, so I for one wouldn't invest in making such a PR given that the feedback focuses on the internal httpx code base look and not the external feature.

**Reply by dctr** [reply]
Being mindful that opinions diverge, I'd like to call out that whether `Optional[T]` looks ugly or not differs from person to person. I myself like it, as I think it makes method signatures more comprehensible, but I can also understand if other people see it differently.

What I would like to focus on here are the practical aspects for users of `httpx`, or a library in general, that should be taken into account:

First, people like myself may write a wrapper around the library, and consequently might pass in `Optional[T]` parameters themselves. In that case the `None` is not a default fallback value "hidden" inside the method if no parameter is passed, but `None` **is** the parameter being passed. So it is not an implicit, but an explicit Optional.

Second, whether an individual developer or team agrees with a particular linting or typing rule or not for their own code base is for them to decide. As a library developer however it is important to cater for a wide range of users, and tighter rules can accompany more loose scenarios, but not the other way around.

Third, @tomchristie, you mentioned you personally dislike the `Optional[T]` syntax, however there is still the alternative of having the respective types being `Union[..., None]` like in the example I provided above. Generally, I am sure there is a way to solve this issue that gets rid of the lint as well as being stylistic to your liking. I personally would argue that this is the exact usage scenario `Optional[T]` was built for, yet both approaches solve the initial problem statement sufficiently.

Lastly, as @altendky pointed out, the likelihood of PR not getting accepted is not a very encouraging premise for someone in the community to help solve this. As mentioned in the third point, there surely could exist a solution that satisfies the maintainers. So I would argue it would be a more healthy approach to facilitate the community in finding a common ground that solves the users' use-case as well as adhering to the maintainers' vision. This will help build the community, encourage others (like myself) to put in the work to create PRs and hopefully ultimately propel the project forward.

**Reply by lovelydinosaur** [reply]
Some good points there. How does this middle ground sound?...

If someone's willing to issue a pull request that *just* changes the signatures in `client.py`, as a starting point, then I'd be willing to take a call on if its worth pushing forward with switching the signatures throughout?

That gives us a visible change against which to judge the trade-offs.

**Reply by dctr** [reply]
Thanks for that @tomchristie , I'll find some time later this week and give that a shot, both with `Optional[T]` as well as `Union[..., None]` and create two PRs for that.

**Reply by dctr** [reply]
Hey @tomchristie , here's a draft PR having a stab at all the types in client.py using the `Union[..., None]` approach.

I chose to try this approach first, as it is already done for some other types and I thought it might have a higher chance of being accepted. However this approach has the clear shortcomming that one is not able to distinguish out when a value is not `None` anymore after a None-check. There is an example in the PR. So this will not solve the problem.

https://github.com/encode/httpx/pull/1760

As pointed out above, `Optional` is the intended tool for this scenario and the issue that surfaced in the PR is a good example for that.

---

There are already some `Optional[T]`s in there, so you can see a bit of how it could look like. I would be up to giving it a shot in another PR if you would be open to consider it. Also, with directly import in types from typing, the code becomes much more concise.

**Reply by lovelydinosaur** [reply]
Okay, in that case I think we should close off #1760, right?

> There are already some Optional[T]s in there, so you can see a bit of how it could look like. I would be up to giving it a shot in another PR if you would be open to consider it. Also, with directly import in types from typing, the code becomes much more concise.

I think the sensible thing here would be to treat each part separately, so firstly *just* issuing a pull request which changes the import style of the `typing` module, without making any other changes.

**Reply by dctr** [reply]
Here we go: https://github.com/encode/httpx/pull/1769

Looks much less noisy now, and I reckon making things `Optional[T]` will also look much cleaner after this.
